text/0000-deref-move.md

```diff
@@ -0,0 +1,322 @@
+case, `Box<T>` has compiler support allowing it to be dereferenced (explicitly
+or implicitly) to produce the inner `T`. Because this is special-cased in the
+compiler, this functionality is not available for other smart pointer types.
+
+Other smart pointers may support this consuming operation, but it must be done
+explicitly due to the lack of language support. For instance, in the stdlib, the
+following methods both consume a smart pointer to produce the inner value:
+
+* `ManuallyDrop<T>::into_inner()`
+* `binary_heap::PeakMut<'a, T>::pop()`
+* `Cow<'a, B>::into_owned()`
+
+I have not made any attempt to survey into the larger ecosystem for more
+examples, but there are probably many more.
+
+Note that removing `FnBox` is explicitly *not* part of the motivation for this
+RFC. It has been suggested in a few places, such as in discussion of
+rust-lang/rust#28796, that `DerefMove` would allow `Box<FnOnce()>` to be called
+without needing the `FnBox` workaround. This is not the case, however: the inner
+value is dynamically-sized and thus cannot be moved onto the stack. Indeed, one
+can call a `Box<T>` where `T` is a concrete type implementing `FnOnce()`.
+`DerefMove` would allow `Box<FnOnce()>` to continue not working, just without a
+special case in the compiler to allow it to do so.
+
+This will not remove all of `Box`'s special cases from the compiler, but
+represents progress towards that goal.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following trait is defined in `std::ops`:
+
+```rust
+trait DerefMove : DerefMut {
+    fn deref_move(self) -> Self::Target;
+}
+```
+
+This trait allows moving out of the result of a derefence. This can happen
+either explicitly, as in `let x = *p;`, or implicitly, as in `p.into_foo()`.
+`Box<T>` from the standard library implements `DerefMove`, allowing you to move
+out of it.
+
+```rust
+let b = Box::new("s".to_owned());
+let s = *b;
+// ERROR: b has already been moved on the previous line.
+// let t = *b;
+```
+
+If the `Target` type is `Copy`, then `DerefMove` will not be used, and instead
+`Deref` will be used and the resulting referenced value will be copied:
+
+```rust
+let b = Box::new(0i32);
+// Since i32 is Copy, this is equivalent to i = b.deref().
+let i = *b;
+// That makes this legal, since b was not moved.
+let j = *b;
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`DerefMove` is a lang item that, like `Deref` and `DerefMut`, plays a role in
+dereferencing, both explicitly with the unary `*` operator and implicitly with
+autoderef.
+
+`deref_move` is only called in contexts where a value is needed to be moved out.
+This can occur either because an explicit dereference is used in a context where
+a value is to be moved, or because autoderef selects a method taking `self` (or
+`Box<self>`) as a parameter. In both cases, when evaluating whether to call
```

Comment:
I forget, is `Box<self>` magic or would this also work for `MyBox<self>`? (should this be made explicit?)

Reply:
`Box<self>` is magic as a receiver type, and I'm not proposing to eliminate that here. It's only mentioned here for completeness.

Reply:
This is sort of getting fixed with arbitrary self types, right?  Or is this specifically about the `Box<self>` syntax?

Reply:
Yes, once arbitrary receiver types are accepted then `Box` won't be magical. But that shouldn't affect this RFC one way or another---it's just that I felt I needed to mention that not all by-value receivers are `Self`. Overload resolution will continue the way it normally is (note: I said "autoderef selects" above, but I really meant overload resolution).